// @markuplint/rules/lib/required-element/index.js
import { createRule } from '@markuplint/ml-core';
import meta from './meta.js';
/**
 * Rule that validates required child elements are present.
 *
 * Accepts an array of CSS selectors as the rule value. For the document level,
 * reports if any matching element is missing from the entire document. For
 * element-level overrides, checks that the specified child elements exist
 * among the element's direct children.
 */
export default createRule({
    meta: meta,
    defaultValue: [],
    defaultOptions: {
        ignoreHasMutableContents: true,
        ignoreOmittedElements: false,
    },
    async verify({ document, report, t }) {
        const hasMutableContent = document.nodeList.some(n => n.is(n.ELEMENT_NODE) && n.hasMutableChildren());
        if (!hasMutableContent) {
            for (const query of document.rule.value) {
                const matched = document.querySelectorAll(query);
                const exists = document.rule.options.ignoreOmittedElements
                    ? [...matched].filter(el => !el.isOmitted)
                    : [...matched];
                if (exists.length === 0) {
                    const message = t('Require {0}', t('the "{0*}" {1}', query, 'element'));
                    report({
                        message,
                        line: 1,
                        col: 1,
                        raw: document.nodeList[0]?.raw.slice(0, 1) ?? '',
                    });
                }
            }
        }
        await document.walkOn('Element', el => {
            if (el.rule.value === document.rule.value) {
                return;
            }
            if (el.rule.options.ignoreHasMutableContents && el.hasMutableChildren()) {
                return;
            }
            for (const query of el.rule.value) {
                const exists = [...el.children].find(child => !(el.rule.options.ignoreOmittedElements && child.isOmitted) && child.matches(query));
                if (!exists) {
                    const message = t('Require {0}', t('the "{0*}" {1}', query, 'element'));
                    report({
                        scope: el,
                        message,
                    });
                }
            }
        });
    },
});
